### PR TITLE
1590: Wordcheck: Be able to reject a suggestion

### PR DESCRIPTION
**Pull request by chrismiceli**
add add words to bad word list
remember cutoff between POSTs

### CODE REVIEW

**Comment by chrismiceli** [comment]
Available in sandbox here:
https://www.pgdp.org/~chrismiceli/c/

Screenshot of new submit button:
image

**Comment by cpeel** [comment]
What you have is good and certainly something that has been requested although I'm not 100% certain it addresses task 1590 as-described. We should probably do the same updates to `tools/project_manager/show_all_good_word_suggestions.php` so the two interfaces are the same (/me rants to self of why those are using different codepaths...).

I think this page could also use a "Reject all suggestions" button, placing it below the "Number of pages with suggestions: X" line. When clicked it would just re-save the good words list. That would reset the this-project-has-suggestions marker and is an oft-requested task. It too would need updating on `show_all_good_word_suggestions.php` for each project -- which would solve Task 1651. The "Reject all suggestions" changes could easily be a separate MR.

**Comment by chrismiceli** [comment]
> 
> 
> What you have is good and certainly something that has been requested although I'm not 100% certain it addresses task 1590 as-described. We should probably do the same updates to `tools/project_manager/show_all_good_word_suggestions.php` so the two interfaces are the same (/me rants to self of why those are using different codepaths...).
> 
> I think this page could also use a "Reject all suggestions" button, placing it below the "Number of pages with suggestions: X" line. When clicked it would just re-save the good words list. That would reset the this-project-has-suggestions marker and is an oft-requested task. It too would need updating on `show_all_good_word_suggestions.php` for each project -- which would solve Task 1651. The "Reject all suggestions" changes could easily be a separate MR.

Great input. I pushed a second commit, but this is still work in progress I think to address all the input in regards to the word check pages. I appreciate the feedback as we work on getting this right.

File: pinc/wordcheck_engine.inc [hunk 1]
```diff
@@ -10,7 +10,7 @@ include_once($relPath.'misc.inc'); // str_contains, array_get
 // NOTE: This should probably be changed when moving to UTF-8
 //       at that time we can probably move to using \W instead of
 //       the mess below
-$word_letters = "a-zA-Z�-��-��-�";
```
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/wordcheck_engine.inc:13]: I suspect your editor saved the file as UTF-8 instead of ISO-8859-1. We actually need the file to stay the latter until the Unicode branch merges (the file was already converted to UTF-8 in that branch).

File: tools/project_manager/show_good_word_suggestions.php [hunk 2]
```diff
@@ -10,6 +10,7 @@
 
 require_login();
 
+define("REJECT_SUGGESTIONS", "reject_suggestions");
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:13]: I would probably put this in `word_freq_table.inc` with the other new defines.

File: tools/project_manager/show_good_word_suggestions.php [hunk 3]
```diff
@@ -37,9 +38,27 @@
...
+        $words = load_project_good_words($projectid);
+        $words = array_merge($words,$postedWords);
+        save_project_good_words($projectid,$words);
+    }
+    else
+    {
+        var_dump("bad word list");
+        $words = load_project_bad_words($projectid);
+        $words = array_merge($words,$postedWords);
+        save_project_bad_words($projectid,$words);
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:60]: Consider moving this outside the if/else block, since it's common to both.

Also, please add spaces after the commas in the function calls (I realize this isn't consistent with the rest of the file).

File: tools/project_manager/show_good_word_suggestions.php [hunk 4]
```diff
@@ -37,9 +38,27 @@
...
+    }
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:61]: Oh hey - here is a fun bug in the old file that is more evident with your code. We determine the last change for the good words file at the top of the file:
```php
$fileObject = get_project_word_file($projectid,"good");
$timeCutoff = get_integer_param($_REQUEST, 'timeCutoff', $fileObject->mod_time, 0, null);
```
But we save the good words list later, changing when it was last updated. This means that after the page is submitted, it will still show the old suggestions instead of the new ones based on the updated time. This is most evident when we "reject all". If we move the above two lines below this `if($format=="update") {` block that should fix it.

File: pinc/wordcheck_engine.inc (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrismiceli) [reply pinc/wordcheck_engine.inc]: Fixed it by changing visual studio code to save in iso-8859-1 for now until unicode comes over.

File: tools/project_manager/show_good_word_suggestions.php (hunk 3 above)
Comment (chrismiceli) [reply tools/project_manager/show_good_word_suggestions.php]: I refactored some, but still needed 2 conditionals for save_project_bad vs save_project_good and load_project_bad vs load_project_good.

File: tools/project_manager/show_good_word_suggestions.php (hunk 4 above)
Comment (chrismiceli) [reply tools/project_manager/show_good_word_suggestions.php:44]: I noticed that and was planning on investigating. This makes perfect sense. Updated.

File: pinc/wordcheck_engine.inc [hunk 5]
```diff
@@ -777,17 +777,6 @@ function save_word_list($path, $words)
     // needed for the array == array comparison below
     $words = array_values($words);
 
-    // see if list has changed before saving it
```
Comment (cpeel) [inline pinc/wordcheck_engine.inc:780]: Instead of taking this out and doing a load/save, let's add a new function to this file that uses `touch()` to update the good word list file. That does less data shuffling and creates a single function we can use where we want to reject suggestions. Perhaps call it simply `touch_project_good_words_file($projectid)`.

File: tools/project_manager/show_good_word_suggestions.php [hunk 6]
```diff
@@ -35,15 +27,33 @@
...
 
+$fileObject = get_project_word_file($projectid,"good");
+$timeCutoff = get_integer_param($_REQUEST, 'timeCutoff', $fileObject->mod_time, 0, null);
+
+if($timeCutoff==0)
+    $time_cutoff_text = _("<b>All proofreader suggestions</b> are included in the results.");
+else
+    $time_cutoff_text = sprintf(_("Only proofreader suggestions made <b>after %s</b> are included in the results."),strftime($datetime_format,$timeCutoff));
+
 list($all_suggestions_w_freq,$all_suggestions_w_occurrences,$round_suggestions_w_freq,$round_suggestions_w_occurrences,$rounds,$round_page_count,$messages) =
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:57]: Here's another fun edge case. If the user uses the new "add to bad list" function to add all suggestions to the bad words list, when the page loads it will say "There are no suggestions in the given time frame that aren't already on the Good Words List." -- which is 100% correct. However, the timestamp of the Good Words List file won't be reset, so the alert on the PM page will still show up. The ideal solution would be to have the alert on the PM listing use the same logic here to determine if there really are good words to show, but that's rather resource intensive. Instead, we could:
```
if($is_reject && $timeCutoff == $fileObject->mod_time && count($all_suggestions_w_occurrences) == 0)
{
    // touch good word list function call
}
```

But trying to implement that in the other file is going to be complicated, and really the whole thing is a mess (which is all my fault 13 years ago).

So for now let's not scope creep the hell out of this task and live with the problem for now.

File: tools/project_manager/word_freq_table.inc [hunk 7]
```diff
@@ -51,6 +55,11 @@ function get_cutoff_script($cutoffOptions,$instances) {
         }
         if(document.getElementById("freqCutoffValue")) document.getElementById("freqCutoffValue").value=newCutoff;
         document.getElementById("current_cutoff").innerHTML=newCutoff;
+
+        // set new value for forms so the value is persisted through actions
+        $('form input[name=freqCutoff').each(function (index, value) {
```
Comment (cpeel) [inline tools/project_manager/word_freq_table.inc:60]: Should this have a `]` in it? i.e. `$('form input[name=freqCutoff]')`?

File: pinc/wordcheck_engine.inc (hunk 5 above)
Comment (chrismiceli) [reply pinc/wordcheck_engine.inc:780]: great idea. I thought of it just after I submitted the last change realizing it is such a waste otherwise.

File: tools/project_manager/word_freq_table.inc (hunk 7 above)
Comment (chrismiceli) [reply tools/project_manager/word_freq_table.inc:60]: Yes, I'm actually very surprised it works without it.

File: tools/project_manager/show_good_word_suggestions.php (hunk 6 above)
Comment (chrismiceli) [reply tools/project_manager/show_good_word_suggestions.php:57]: Wouldn't the touch get called during the is_reject update processing?

Comment (cpeel) [reply tools/project_manager/show_good_word_suggestions.php:57]: Not in the case where they add the last word to the bad words list. It would if they rejected all the suggestions,

Comment (chrismiceli) [reply tools/project_manager/show_good_word_suggestions.php:57]: I see, I was getting confused between "add to bad list" and reject all suggestions.

File: tools/project_manager/show_good_word_suggestions.php [hunk 8]
```diff
@@ -15,38 +15,58 @@
...
     _get_word_list($projectid,$timeCutoff);
 
+if ($destination_list == BAD_WORD_LIST && $timeCutoff == $fileObject->mod_time && $all_suggestions_w_occurrences == 0)
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:65]: You're missing a `count()` on the array:
```
if ($destination_list == BAD_WORD_LIST && $timeCutoff == $fileObject->mod_time && count($all_suggestions_w_occurrences) == 0)
```

Comment (cpeel) [reply tools/project_manager/show_good_word_suggestions.php:65]: The following diff will do effectively the same thing for the `show_all_good_word_suggestions.php` file:
```diff
diff --git a/tools/project_manager/show_all_good_word_suggestions.php b/tools/project_manager/show_all_good_word_suggestions.php
index 34fc53f..db5d9e4 100644
--- a/tools/project_manager/show_all_good_word_suggestions.php
+++ b/tools/project_manager/show_all_good_word_suggestions.php
@@ -186,7 +186,17 @@ echo "<br>";

         // if no words are returned (probably because something was
         // suggested but is no longer in the text) skip this project
-        if(count($suggestions_w_freq)==0) continue;
+        if(count($suggestions_w_freq)==0)
+        {
+            // If the time cutoff is the same as the good words file and
+            // we still have no suggestions, touch the good words file to
+            // reset the time so no alert appears for this project.
+            if($timeCutoffActual == $goodFileObject->mod_time)
+            {
+                touch_project_good_words($projectid);
+            }
+            continue;
+        }

         $projectsNeedingAttention++;
```

File: pinc/wordcheck_engine.inc [hunk 9]
```diff
@@ -777,7 +783,7 @@ function save_word_list($path, $words)
...
+     // see if list has changed before saving it
```
Comment (cpeel) [inline pinc/wordcheck_engine.inc:786]: Minor: extra space here (I assume from putting this back in).

File: tools/project_manager/show_all_good_word_suggestions.php [hunk 10]
```diff
@@ -79,7 +99,9 @@
     // get all projects for this PM
     $projects = _get_projects_for_pm($pm);
 
+    $rejectAlllabel = _("Reject all suggestions");
     $submitLabel = _("Add selected words to Good Words List");
+    $rejectLabel = _("Add selected words to Bad Words List");
```
Comment (cpeel) [inline tools/project_manager/show_all_good_word_suggestions.php:104]: Consider changing this to `$submitBadLabel` instead, since putting them on the bad list isn't rejecting them.

File: tools/project_manager/show_good_word_suggestions.php [hunk 11]
```diff
@@ -151,6 +171,7 @@
 echo_cutoff_text( $initialFreq,$cutoffOptions );
 
 $submit_label=_("Add selected words to Good Words List");
+$reject_label = _("Add selected words to Bad Words List");
```
Comment (cpeel) [inline tools/project_manager/show_good_word_suggestions.php:174]: Suggest changing this to `$submit_bad_label` instead.

File: tools/project_manager/show_project_possible_bad_words.php [hunk 12]
```diff
@@ -107,11 +107,11 @@
 $checkbox_form["freqCutoff"]=$freqCutoff;
 echo_checkbox_selects(count($bad_words_w_freq));
 echo_checkbox_form_start($checkbox_form);
-echo_checkbox_form_submit(_("Add selected words to Bad Words List"));
+echo_checkbox_form_submit(_("Add selected words to Bad Words List"), BAD_WORD_LIST);
```
Comment (srjfoo) [inline tools/project_manager/show_project_possible_bad_words.php:110]: The possibility exists that a PM might want to add a word to the GWL that is identified as a possible bad word. If I looked at the following, I'd just as soon add "tip" to the GWL and not bother the proofers with worrying about it.

image

and 

image

I realize that the task really deals only with the GWL, and this could be considered scope creep -- should it be a different task?

Comment (cpeel) [reply tools/project_manager/show_project_possible_bad_words.php:110]: Yes, that's scope creep and we should punt that. I've already scope-creeped this a bit much already for poor Chris.

File: tools/project_manager/show_all_good_word_suggestions.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply tools/project_manager/show_all_good_word_suggestions.php:104]: > Am I correct that "Reject all suggestions" is just a means of clearing the flag so that a PM knows they've finished with this word list? That is, the words that were rejected wholesale are not on a list to be used as a filter for future suggestions and never seen again?

Correct.

[the remark this answers is not in the document]
Comment (srjfoo) [reply tools/project_manager/show_all_good_word_suggestions.php:104]: Thanks for the confirmation. Either way, it needs to be documented exactly how it works, so there's no misunderstanding. Possibly as part of the recent additions announcement.

File: tools/project_manager/show_project_possible_bad_words.php (hunk 12 above)
Comment (srjfoo) [reply tools/project_manager/show_project_possible_bad_words.php:110]: <grin> Understood.

It should be documented, then, and perhaps a new task created after it's rolled out. I could always be wrong, but this strikes me as the sort of inconsistency that will be noticed and questioned.

File: tools/project_manager/show_all_good_word_suggestions.php [hunk 13]
```diff
@@ -11,6 +11,9 @@
 
 require_login();
 
+define("GOOD_WORD_LIST", "good_word_list");
+define("BAD_WORD_LIST", "bad_word_list");
```
[the remark this answers is not in the document]
Comment (chrismiceli) [reply tools/project_manager/show_all_good_word_suggestions.php]: some of the code has it plural, others don't. I must've most recently seen one singular and went with that. I'll update.

File: tools/project_manager/show_project_possible_bad_words.php (hunk 12 above)
Comment (chrismiceli) [reply tools/project_manager/show_project_possible_bad_words.php:110]: Create the ticket. This shouldn't be hard to add. Just trying to not go past the upcoming code freeze.

**Comment by cpeel** [comment]
Right now, if a PM rejects words we still show the words they rejected in the form because the `timeCutoff` submitted with the form is the old GWL save time. This patch fixes that such that if the user is looking at the "since last saved time", they will be after submitting the form:
```patch
diff --git a/tools/project_manager/show_good_word_suggestions.php b/tools/project_manager/show_good_word_suggestions.php
index c6195bc..bba5bc2 100644
--- a/tools/project_manager/show_good_word_suggestions.php
+++ b/tools/project_manager/show_good_word_suggestions.php
@@ -175,7 +175,7 @@ $submit_bad_label = _("Add selected words to Bad Words List");

 $checkbox_form["projectid"]=$projectid;
 $checkbox_form["freqCutoff"]=$freqCutoff;
-$checkbox_form["timeCutoff"]=$timeCutoff;
+$checkbox_form["timeCutoff"]=$timeCutoff == $fileObject->mod_time ? "" : $timeCutoff;
 echo_checkbox_form_start($checkbox_form);

 $context_array["[[TITLE]]"]=_("Show Context");
diff --git a/tools/project_manager/word_freq_table.inc b/tools/project_manager/word_freq_table.inc
index 3782071..6dd9221 100644
--- a/tools/project_manager/word_freq_table.inc
+++ b/tools/project_manager/word_freq_table.inc
@@ -423,7 +423,8 @@ function echo_checkbox_form_start($checkbox_form) {
     echo "<form action='" . attr_safe($_SERVER['PHP_SELF']) . "' method='post'>";
     echo "<input type='hidden' name='format' value='update'>";
     foreach($checkbox_form as $name => $value) {
-        echo "<input type='hidden' name='$name' value='$value'>";
+        if($value != '')
+            echo "<input type='hidden' name='$name' value='$value'>";
     }
 }
```

**Comment by chrismiceli** [comment]
Sandbox updated to latest commit

**Comment by cpeel** [comment]
This works great in all of the variations that I tested!
